lib/auth/version.go

```diff
@@ -20,55 +20,101 @@ package auth
 
 import (
 	"context"
+	"log/slog"
 	"os"
 
 	"github.com/coreos/go-semver/semver"
 	"github.com/gravitational/trace"
+
+	authinfov1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/authinfo/v1"
+	"github.com/gravitational/teleport/api/types/authinfo"
+	"github.com/gravitational/teleport/lib/services"
 )
 
 const (
 	// skipVersionUpgradeCheckEnv is environment variable key for disabling the check
 	// major version upgrade check.
 	skipVersionUpgradeCheckEnv = "TELEPORT_UNSTABLE_SKIP_VERSION_UPGRADE_CHECK"
+	// versionUpgradeCheckMaxWriteRetry is the number of retries for conditional updates of the AuthInfo resource.
+	versionUpgradeCheckMaxWriteRetry = 5
 )
 
 // validateAndUpdateTeleportVersion validates that the major version persistent in the backend
 // meets our upgrade compatibility guide.
 func validateAndUpdateTeleportVersion(
 	ctx context.Context,
-	storage VersionStorage,
-	currentVersion *semver.Version,
-) error {
+	procStorage VersionStorage,
+	backendStorage services.AuthInfoService,
+	currentVersion semver.Version,
+) (err error) {
 	if skip := os.Getenv(skipVersionUpgradeCheckEnv); skip != "" {
 		return nil
 	}
 
-	lastKnownVersion, err := storage.GetTeleportVersion(ctx)
-	if trace.IsNotFound(err) {
-		if err := storage.WriteTeleportVersion(ctx, currentVersion); err != nil {
+	for range versionUpgradeCheckMaxWriteRetry {
+		authInfo, err := backendStorage.GetAuthInfo(ctx)
+		if trace.IsNotFound(err) {
+			// TODO(vapopov): DELETE IN v19.0.0 – the last known version should already be migrated to backend storage.
+			// Fallback to local process storage for backward compatibility with previous versions.
+			lastKnownVersion, err := procStorage.GetTeleportVersion(ctx)
+			if trace.IsNotFound(err) {
+				authInfo, err = authinfo.NewAuthInfo(&authinfov1.AuthInfoSpec{TeleportVersion: currentVersion.String()})
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				if _, err := backendStorage.CreateAuthInfo(ctx, authInfo); err != nil {
+					err = trace.Wrap(err)
+					slog.WarnContext(ctx, "Failed to create AuthInfo resource",
+						"error", err)
+					continue
+				}
+				return nil
+			} else if err != nil {
+				return trace.Wrap(err)
+			}
+
+			authInfo, err = authinfo.NewAuthInfo(&authinfov1.AuthInfoSpec{TeleportVersion: lastKnownVersion.String()})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if _, err := backendStorage.CreateAuthInfo(ctx, authInfo); err != nil {
```

Comment:
What's the point of creating the `auth_info` with the stored version just to immediately write it again?

I would consider moving the migration part of this out of the loop and maybe have it work in the opposite way: if there's a version in process storage then create it in the backend (ignoring AlreadyExists) and then delete the version from process storage, then proceed with the regular logic.

Reply:
if there many auth servers we are going to continuously override the version in backend until every instance drop their local last known version